### PR TITLE
CLI for transforming output

### PR DESCRIPTION
**Pull request by oschaaf**
Allows transformations to different output formats. 
This will become more useful as the number of supported output formats grows. 
(Dotted strings and fortio are both expected / in the pipeline).

This first cut support `stdin` for taking input, expecting json.

Example I/O, transforming `json` to `human` output:

```
bazel-bin/nighthawk_client --duration 1 --output-format json 127.0.0.1 | bazel-bin/nighthawk_output_transform --output-format human
[11:48:47.109362][16175][I] [source/client/process_impl.cc:171] Starting 1 threads / event loops. Test duration: 1 seconds.
[11:48:47.109442][16175][I] [source/client/process_impl.cc:173] Global targets: 1 connections and 5 calls per second.
[11:48:47.121150][16175][I] [external/envoy/source/common/upstream/cluster_manager_impl.cc:162] cm init: all clusters initialized
[11:48:49.051875][16175][I] [source/client/client.cc:56] Done.
Nighthawk - A layer 7 protocol benchmarking tool.

Queueing and connection setup latency
  samples: 4
  mean:    0s 000ms 039us
  pstdev:  0s 000ms 025us

  Percentile  Count       Latency        
  0           1           0s 000ms 019us 
  0.5         2           0s 000ms 026us 
  0.75        3           0s 000ms 028us 
  1           4           0s 000ms 082us 

Request start to response end
  samples: 4
  mean:    0s 000ms 334us
  pstdev:  0s 000ms 196us

  Percentile  Count       Latency        
  0           1           0s 000ms 202us 
  0.5         2           0s 000ms 224us 
  0.75        3           0s 000ms 236us 
  1           4           0s 000ms 673us 

Initiation to completion
  samples: 4
  mean:    0s 000ms 423us
  pstdev:  0s 000ms 249us

  Percentile  Count       Latency        
  0           1           0s 000ms 249us 
  0.5         2           0s 000ms 290us 
  0.75        3           0s 000ms 299us 
  1           4           0s 000ms 854us 

Counter                                 Value       Per second
benchmark.http_2xx                      5           5.00
cluster_manager.cluster_added           1           1.00
membership_change                       1           1.00
runtime.load_success                    1           1.00
runtime.override_dir_not_exists         1           1.00
upstream_cx_destroy                     1           1.00
upstream_cx_destroy_local               1           1.00
upstream_cx_http1_total                 1           1.00
upstream_cx_rx_bytes_total              55840       55840.00
upstream_cx_total                       1           1.00
upstream_cx_tx_bytes_total              285         285.00
upstream_rq_pending_total               1           1.00
upstream_rq_total                       5           5.00

```

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
Cross referencing tracking issue #151
Also, /cc @nareddyt

**Comment by oschaaf** [comment]
(Had to restart CI, clang-tidy was flaking. It looks like somehow memory usage became a problem again; we may want to watch this because if it happens again lowering parallelism in CI is probably warranted)

File: source/client/output_transform_main.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    : input_(input) {
+  const char* descr = "L7 (HTTP/HTTPS/HTTP2) performance characterization transformation tool.";
+  TCLAP::CmdLine cmd(descr, ' ', "PoC"); // NOLINT
+
+  std::vector<std::string> output_formats = {"human", "yaml", "json"};
+  TCLAP::ValuesConstraint<std::string> output_formats_allowed(output_formats);
+  output_format_ = "";
+  TCLAP::ValueArg<std::string> output_format(
+      "", "output-format",
+      fmt::format("Verbosity of the output. Possible values: [human, yaml, json]."), true, "",
```
Comment (htuch) [inline source/client/output_transform_main.cc]: Is this verbosity or format?

File: source/client/output_transform_main.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    return 1;
+  }
+
+  // We override the output format, which will make the OutputCollectorFactory hand us
+  // an instance that will output the desired format for us.
+  output.mutable_options()->mutable_output_format()->set_value(translated_format);
+  OutputCollectorPtr collector;
+  try {
+    const auto options = OptionsImpl(output.options());
+    OutputCollectorFactoryImpl factory(time_system_, options);
```
Comment (htuch) [inline source/client/output_transform_main.cc]: This PR looks like the right direction for sure. I'm a bit confused by the appearance of "output collector" in a reformatting tool though; ideally we have a library that is just an output formatter, which maps from proto/JSON to N formats. This shouldn't have any relationship to collecting output as we previously had this code.

File: source/client/output_transform_main.cc (hunk 1 above)
Comment (oschaaf) [reply source/client/output_transform_main.cc]: Whoops, fixed

Comment (oschaaf) [reply source/client/output_transform_main.cc]: Also, added a way to pull the supported formats as a string vector in a canonical fashion.

File: source/client/output_transform_main.cc (hunk 2 above)
Comment (oschaaf) [reply source/client/output_transform_main.cc]: I refactored, and extracted a new `OutputFormatter` from `OutputCollector`.

File: include/nighthawk/client/output_formatter.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+namespace Nighthawk {
+namespace Client {
+
+/**
+ * Transforms from nighthawk::client::Output to an output format.
+ */
+class OutputFormatter {
+public:
+  virtual ~OutputFormatter() = default;
+  virtual void setProto(const nighthawk::client::Output& output) PURE;
```
Comment (htuch) [inline include/nighthawk/client/output_formatter.h]: Why not just have a `std::string formatProto(proto)` method?

File: source/client/output_transform_main.cc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  output_format_ = output_format.getValue();
+}
+
+std::string OutputTransformMain::readInput() {
+  std::stringstream input;
+  std::string line;
+  while (getline(input_, line)) {
+    input << line << std::endl;
+  }
+  return input.str();
```
Comment (htuch) [inline source/client/output_transform_main.cc]: There has to be a better way to do this in C++ I think, see how Envoy does it in https://github.com/envoyproxy/envoy/blob/master/source/common/filesystem/posix/filesystem_impl.cc#L92.

File: include/nighthawk/client/output_formatter.h (hunk 3 above)
Comment (oschaaf) [reply include/nighthawk/client/output_formatter.h]: good point, done.

File: source/client/output_transform_main.cc (hunk 4 above)
Comment (oschaaf) [reply source/client/output_transform_main.cc]: adjusted to use `rdbuf()`

**Comment by oschaaf** [comment]
Feedback addressed in https://github.com/envoyproxy/nighthawk/pull/159/commits/0043c6907b6668f35c9efc97276035ea1464a673, thanks!

/assign htuch

**Pull request merged**